src/metabase/sync_database/cached_values.clj

```diff
@@ -0,0 +1,157 @@
+         (<= total-length field-values-total-max-length))))
+
+(defn test-for-cardinality?
+  "Should FIELD should be tested for cardinality?"
+  [{:keys [base_type] :as field}]
+  (or (field-values/field-should-have-field-values? field)
+      (and #_(nil? (:special_type field))
+           #_is-new? ;; do we actually want to test for this here?
+           (not (isa? base_type :type/DateTime))
+           (not (isa? base_type :type/Collection))
+           (not (= base_type :type/*)))))
+
+;; was part of test:cardinality-extract-field-values
+(defn extract-field-values
+  "Extract field-values for FIELD.  If number of values exceeds `low-cardinality-threshold` then we return an empty set of values."
+  [{:keys [name base_type] :as field} field-stats]
+  ;; TODO: we need some way of marking a field as not allowing field-values so that we can skip this work if it's not appropriate
+  ;;       for example, :type/Category fields with more than MAX values don't need to be rescanned all the time
+
+  (let [#_name-type-guess #_(infer-special-type/infer-field-special-type name base_type)
+        ;collecting-field-values-is-allowed? true (field-values/field-should-have-field-values? field)
+;        _ (log/errorf (u/format-color 'green "name: %s :type %s" name name-type-guess))
+        non-nil-values  (when #_(test-for-cardinality? field) #_(nil? name-type-guess) (field-values/field-should-have-field-values? field)
+                          (filter identity (queries/field-distinct-values field (inc classify/low-cardinality-threshold))))
+        ;; only return the list if we didn't exceed our MAX values and if the the total character count of our values is reasable (#2332)
+        distinct-values (when (field-values-below-low-cardinality-threshold? non-nil-values)
+                          non-nil-values)]
+    (if (seq distinct-values)
+      (assoc field-stats :values distinct-values)
+      field-stats)))
+
+(defn make-analyze-table
+  "Make a generic implementation of `analyze-table`."
+  {:style/indent 1}
+  [driver & {:keys [calculate-row-count?]
+             :or   {calculate-row-count?  true}}]
+  (fn [driver table new-field-ids]
+    {#_:row_count #_(when calculate-row-count? (u/try-apply table-row-count table))
+     :fields    (for [id new-field-ids]
+                  (extract-field-values (field/Field id) {:id id}))}))
+
+(defn generic-analyze-table
+  "An implementation of `analyze-table` using the defaults (`default-field-avg-length` and `field-percent-urls`)."
+  [driver table new-field-ids]
+  ((make-analyze-table driver) driver table new-field-ids))
+
+;;;; End of driver stuff
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;;; call drivers, get field values, save them in application DB
+(defn cache-table-data-shape!
```

Comment:
We should also have a function to cache the values for a single field. It would be really useful if it accepted a flag for a sampled vs full scan.